### PR TITLE
Add Let's Encrypt HTTP Challenge

### LINKED ISSUE #1832: Feature: http-01 support
I'm trying to deploy traefik behind a reverse proxy that doesn't support tls-sni-01, instead I want to use a challenge that works over http. In the code it appears traefik currently only supports explicitly dns-01 and tls-sni-01. Is this correct? For my setup to work, I need http-01. Sadly I couldn't find anything related in the issue tracker.

### What did you do?
Configure traefik to acquire certificates with acme.

### What did you expect to see?
Valid certificate from let's encrypt

### What did you see instead?
traefik default cert

### Output of `traefik version`:
```
Version: v1.3.2
Codename: raclette
Go version: go1.8.3
Built: 2017-06-29_04:52:35PM
OS/Arch: linux/amd64
```

### PR DESCRIPTION
**Pull request by Juliens**
### What does this PR do?

Until now, Træfik only allowed `TLS-SNI-01` and `DNS-01` challenges to generate/renew ACME certificates.
This PR allows users to generate/renew ACME certificates thanks to the `HTTP-01` challenge.

### Motivation

Due to a security vulnerability, Let's Encrypt disabled `TLS-SNI-01` challenges.
More information [here](https://community.letsencrypt.org/t/2018-01-09-issue-with-tls-sni-01-and-shared-hosting-infrastructure/49996) and [here](https://community.letsencrypt.org/t/2018-01-11-update-regarding-acme-tls-sni-and-shared-hosting-infrastructure/50188).

Fixes #1832 

### More

- [x] Added/updated tests
- [x] Added/updated documentation

### Additional Notes

* For now, `TLS-SNI-01` stays the default challenge in Træfik. If this challenge becomes definitively disabled, it will be removed from Træfik in future versions.
* `HTTP-01` challenge requires an `HTTP` entry point which can be accessed through the port `80` by Let's Encrypt.
* The `acme` configuration has to evolve in Træfik, many fields will be deprecated by this PR.

### CODE REVIEW

File: acme/acme.go [hunk 1]
```diff
@@ -33,25 +36,37 @@ var (
...
+	challengeTLSProvider  *challengeTLSProvider
+	challengeHTTPProvider *challengeHTTPProvider
+	checkOnDemandDomain   func(domain string) bool
+	jobs                  *channels.InfiniteChannel
+	TLSConfig             *tls.Config `description:"TLS config in case wildcard certs are used"`
+	dynamicCerts          *safe.Safe
+}
+
+type DNSChallenge struct {
+	DNSProvider       string `description:"Use a DNS based challenge provider rather than HTTPS."`
```
Comment (emilevauge) [inline acme/acme.go]: s/DNSProvider/Provider

File: acme/acme.go [hunk 2]
```diff
@@ -33,25 +36,37 @@ var (
...
+	DelayDontCheckDNS int    `description:"Assume DNS propagates after a delay in seconds rather than finding and querying nameservers."`
+}
+
+type HTTPChallenge struct {
+	EntryPoint string `description: "HTTP EntryPoint for challenge to"`
```
Comment (emilevauge) [inline acme/acme.go]: `HTTP EntryPoint for challenge to` => `HTTP challenge EntryPoint`

File: acme/acme.go [hunk 3]
```diff
@@ -112,10 +127,36 @@ func (a *ACME) init() error {
...
+
 	a.jobs = channels.NewInfiniteChannel()
 	return nil
 }
 
+func (a *ACME) AddRoutes(router *mux.Router) {
+	router.Methods(http.MethodGet).Path(acme.HTTP01ChallengePath("{token}")).Handler(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		vars := mux.Vars(req)
+		if token, ok := vars["token"]; ok {
+			domain, _, _ := net.SplitHostPort(req.Host)
```
Comment (emilevauge) [inline acme/acme.go]: `err` should not be ignored

File: acme/account.go [hunk 4]
```diff
@@ -24,6 +24,7 @@ type Account struct {
 	PrivateKey         []byte
 	DomainsCertificate DomainsCertificates
 	ChallengeCerts     map[string]*ChallengeCert
+	HttpChallenge      map[string]map[string][]byte
```
Comment (emilevauge) [inline acme/account.go]: We really need to add an integration test to check that we support old acme.json.

File: acme/acme.go [hunk 5]
```diff
@@ -155,7 +196,8 @@ func (a *ACME) CreateClusterConfig(leadership *cluster.Leadership, tlsConfig *tl
 	}
 
 	a.store = datastore
-	a.challengeProvider = &challengeProvider{store: a.store}
+	a.challengeTLSProvider = &challengeTLSProvider{store: a.store}
+	a.challengeHTTPProvider = &challengeHTTPProvider{store: a.store}
```
Comment (emilevauge) [inline acme/acme.go]: I'm not in love with instantiating `challengeTLSProvider` & `challengeHTTPProvider` here.

File: acme/acme.go [hunk 6]
```diff
@@ -112,10 +127,36 @@ func (a *ACME) init() error {
...
+
+	if len(a.DNSProvider) > 0 {
+		log.Warn("ACME.DNSProvider is deprecated, use ACME.DNSChallenge instead")
+		a.DNSChallenge = &DNSChallenge{DNSProvider: a.DNSProvider, DelayDontCheckDNS: a.DelayDontCheckDNS}
+	}
+
+	if a.OnDemand {
+		log.Warn("ACME.OnDemand is deprecated")
+	}
+
```
Comment (emilevauge) [inline acme/acme.go]: I would use `SetEffectiveConfiguration()` to initialize the configuration correctly instead of `init()`

**Comment by dfeyer** [comment]
Based on https://community.letsencrypt.org/t/2018-01-11-update-regarding-acme-tls-sni-and-shared-hosting-infrastructure/50188 I think you can change the default ...

And a huge thanks for your reactivity to implement HTTP-01

File: docs/user-guide/examples.md [hunk 7]
```diff
@@ -153,10 +172,11 @@ This configuration allows generating a Let's Encrypt certificate during the firs
 [acme]
 email = "test@traefik.io"
 storage = "acme.json"
-dnsProvider = "digitalocean" # DNS Provider name (cloudflare, OVH, gandi...)
-delayDontCheckDNS = 0
 caServer = "http://172.18.0.1:4000/directory"
 entryPoint = "https"
+  [acme.dsnChallenge]
```
Comment (Luit) [inline docs/user-guide/examples.md]: typo

**Comment by sps-aboalarm** [comment]
This fix does not work for me. I get "Error while ACME challenge: address test.test.at: missing port in address"

I took a quick look into the code and think: In acme.go                        

domain, _, err := net.SplitHostPort(req.Host)

is getting only the host without port and failing. So for now i'm stuck. I have no idea, why this is not working for me, but obviously works for others. My setup is quite basic: traefik on docker with no special thinks (i reduced the config mostly to the basic example).

**Comment by markuslackner** [comment]
This fix does not work for me. I get "Error while ACME challenge: address test.test.at: missing port in address". I took a quick look into the code and think: In acme.go 

domain, _, err := net.SplitHostPort(req.Host) 

is getting only the host without port and failing. So for now i'm stuck. I have no idea, why this is not working for me, but obviously works for others. My setup is quite basic: traefik on docker with no special thinks (i reduced the config mostly to the basic example).

**Comment by ldez** [comment]
@markuslackner Could you go to the [Traefik community Slack channel](https://traefik.herokuapp.com) in the channel #support.

**Comment by jmagnusson** [comment]
Is there perhaps any chance that we could see an rc5 release to dockerhub today, with this fix included?

File: docs/configuration/acme.md [hunk 8]
```diff
@@ -78,26 +85,64 @@ entryPoint = "https"
...
 # [[acme.domains]]
-# main = "local3.com"
+#   main = "local3.com"
 # [[acme.domains]]
-# main = "local4.com"
+#   main = "local4.com"
+
+# Use a HTTP-01 acme challenge rather than TLS-SNI-01 challenge
+#
+# Optional but advisable
```
Comment (emilevauge) [inline docs/configuration/acme.md]: s/advisable/recommend

File: docs/user-guide/cluster-docker-consul.md [hunk 9]
```diff
@@ -56,15 +56,18 @@ $ traefik \
...
 ```
 
-Let's Encrypt needs 3 parameters: an entry point to listen to, a storage for certificates, and an email for the registration.
+Let's Encrypt needs 4 parameters: an TLS entry point to listen to, a non-TLS entry point to allow HTTP challenges, a storage for certificates, and an email for the registration.
 
 To enable Let's Encrypt support, you need to add `--acme` flag.
 
 Now, Træfik needs to know where to store the certificates, we can choose between a key in a Key-Value store, or a file path: `--acme.storage=my/key` or `--acme.storage=/path/to/acme.json`.
 
+The `acme.httpChallenge.entryPoint` flag enables the `HTTP-01` challenge and specifies the entryPoint to use for the challenges.
```
Comment (emilevauge) [inline docs/user-guide/cluster-docker-consul.md]: s/to use for the challenges/to use during the challenges

**Comment by sebasjm** [comment]
Currently using this release. Thanks!
I have build this release and manually pushed here `sebasjm/traefik:v1.5.0-rc5`, if anyone want to use it for testing purpose. Wait the official image for production.

**Comment by Rzeszow** [comment]
Or use from here: https://hub.docker.com/r/containous/traefik/ 
`docker pull containous/traefik:latest`

**Comment by kruemelro** [comment]
I'm still getting the 
`acme: Could not determine solvers`
Error with the latest Version. Did i've overseen something?

**Comment by Rzeszow** [comment]
@kruemelro Remember to download a new image and clear the cache.
Do not forget to add:
```
[acme.httpChallenge]
entryPoint = "http"
```
to config file (`traefik.toml`)

**Comment by ldez** [comment]
The documentation http://v1-5.archive.docs.traefik.io/configuration/acme/#acmehttpchallenge

Please don't use `containous/traefik:latest`

**Comment by Rzeszow** [comment]
@ldez Currently it is the only way because Docker has not yet released a new version of the official image.

**Comment by ldez** [comment]
I know but we don't recommend the use of this image.

**Comment by ldez** [comment]
The official image is available: `docker pull traefik:v1.5.0-rc5`

**Comment by HUg0005** [comment]
@ldez When I try run `docker pull traefik:v1.5.0-rc5` on my raspberry pi I get `ERROR: no matching manifest for linux/arm in the manifest list entries`.

**Comment by ldez** [comment]
@HUg0005 I think some images are still under construction, you need to wait.
If you always have this problem in 3-4 hours, come on the [Traefik community Slack](https://traefik.herokuapp.com) to inform us.

**Comment by xalexslx** [comment]
Hi,

I'm following this PR since the beginning. Thanks all for the efforts. Can I still use the redirection to https entrypoint enabled on the http entrypoint or it won't work properly with the HTTP-01 validation?

If no, how can I redirect anything besides the validation to https entrypoint?

Thanks.

**Comment by ldez** [comment]
@xalexslx you can still use the redirection.

**Comment by signaleleven** [comment]
Just rebuilt my container with traefik:1.5 and the new acme config: works like a charm. 
THANKS to everyone involved, it's been a pleasure seeing the project reacting to this. I'm humbled by such awesomeness. You guys rock.

**Comment by vrosales** [comment]
Hi, i am getting 400/timeout when enabling HTTP-01, is there anything else i am supposed to do beside enabling the challenge?

    time="2018-01-16T13:28:28Z" level=error msg="Error getting ACME certificates [staging2.example.com] : cannot obtain certificates map[staging2.example.com:acme: Error 400 - urn:acme:error:connection - Fetching http://staging2.example.com/.well-known/acme-challenge/Gw8J6YHJRYT6Tokf3moo8WzzdHuOXe9qcofZ38KuPpk: Timeout
    Error Detail:
	    Validation for staging2.example.com:80
	    Resolved to:
		    XX.XX.XX.XX
	    Used: XX.XX.XX.XX

**Comment by nmengin** [comment]
@vrosales Can you  come on the [Traefik community Slack](https://traefik.herokuapp.com) to give us more information about your configuration?

**Comment by ldez** [comment]
Due to the number of question about redirection:

The redirection is fully compatible with the HTTP-01 challenge.

You can use redirection with HTTP-01 challenge without problem.

Enjoy 🎉

**Pull request merged**